content/en/docs/Writing policies/validate.md

```diff
@@ -460,3 +460,58 @@ spec:
       message: "Changing default network policies is not allowed."
       deny: {}
 ```
+
+## foreach
+
+The `foreach` declaration simplifies validation of sub-elements in resource declarations, for example Containers in a Pod. 
+
+A `foreach` must contain a `list` attribute that defines the list of elements it processes. For example, iterating over the list of containers in a Pod is performed using this `list` declaration:
+
+```yaml
+list: request.object.spec.containers
+```
+
+When a `foreach` is processed, the Kyverno engine will expand fetch the list of elements specified by `list`.
+
+A variable `element` is added to the processing context on each interation. This allows referencing data in the element using `element.<name>` where name is the attribute name. For example, using the list `request.object.spec.containers` when the `request.object` is a Pod allows referencing the container image as `element.name` withing a `foreach`.
+
+The following child declarations are permitted in a `foreach`:
+- [Patterns](/docs/writing-policies/validate/#patterns)
+- [AnyPatterns](/docs/writing-policies/validate/#anypattern)
+- [Deny](/docs/writing-policies/validate/#deny-rules)
+
+
+In addition, each `foreach` declaration can contain the following declarations:
+- [Context](https://kyverno.io/docs/writing-policies/external-data-sources/): to add additional external data only available per loop iteration. 
+- [Preconditions](/docs/writing-policies/preconditions/): to control when a loop iteration is skipped
+
+Here is a complete example to enforce that all container images are from a trusted registry:
+
+```yaml
+apiVersion : kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: check-images
+spec:
+  validationFailureAction: enforce
+  background: false
+  rules:
+  - name: check-registry
+    match:
+      resources:
+        kinds:
+        - Pod
+    preconditions:
+      - key: "{{request.type}}"
+        operator: NotEquals
+        value: "DELETE"
+    validate:
+      message: "unknown registry"  
+      foreach:
+        # NOTE: this does not handle initContainer. This limitation will be  
+        # addressed by https://github.com/kyverno/kyverno/issues/2505.
+        list: "request.object.spec.containers"
+        pattern:
+          image: "trusted-registry.io/*"
```

Comment:
We don't need to specify `.spec.containers` under pattern? For example:
```yaml
        list: "request.object.spec.containers"
        pattern:
          spec:
            template:
              image: ...
```

It seems different from mutate foreach.

Reply:
What does the auto-gen rule look like?

Reply:
When using a pattern, it is applied to the `element`. Hence,  `spec.containers` is not required.

However, in the `patchStrategicMerge` the patch is currently applied to the `request.object`. I left it this way as I think we reference the schema of the object. Or can we apply a  `patchStrategicMerge` to a `element`?